File: 29-divide-two-integers/29-divide-two-integers.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        isNegative = (dividend<0 and divisor>0) or (dividend>0 and divisor<0)
        res=0;
        dividend,divisor=abs(dividend),abs(divisor);
        
        while(dividend >= divisor):
            curr_divisor,num_divisor=divisor,1;
            while(dividend>=curr_divisor):
                dividend-=curr_divisor;
                res+=num_divisor;
                curr_divisor<<=1;
                num_divisor<<=1;
                
        if(isNegative):
            res=-res;
            
        return  min(max(-2147483648, res), 2147483647)

File: 29-divide-two-integers/29-divide-two-integers.cpp
class Solution {
public:
     long divide(long dividend, long divisor) {
        if (dividend == INT_MIN && divisor == -1) {
            return INT_MAX;
        }
        int sign = (dividend > 0) ^ (divisor > 0) ? -1 : 1;
        dividend = labs(dividend), divisor = labs(divisor);
        long ans=0;
        while (dividend >= divisor) {
            long curr_divisor = divisor, num_divisor = 1;
            while ((curr_divisor << 1) <= dividend) {
                curr_divisor <<= 1;
                num_divisor <<= 1;
            }
            dividend -= curr_divisor;
            ans +=num_divisor;
        }
        return sign*ans;
    }
};